File: app/src/main/java/openfoodfacts/github/scrachx/openfood/views/FullScreenActivityOpener.java
package openfoodfacts.github.scrachx.openfood.views;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v4.app.Fragment;
import android.view.View;
import openfoodfacts.github.scrachx.openfood.R;
import openfoodfacts.github.scrachx.openfood.images.ImageKeyHelper;
import openfoodfacts.github.scrachx.openfood.models.Product;
import openfoodfacts.github.scrachx.openfood.models.ProductImageField;
import openfoodfacts.github.scrachx.openfood.utils.LocaleHelper;
import org.apache.commons.lang3.StringUtils;

/**
 * Used to open fullscreen activity
 */
public class FullScreenActivityOpener {
    private FullScreenActivityOpener() {

    }

    public static void openForUrl(Fragment fragment, Product product, ProductImageField imageType, String mUrlImage, View mImageFront) {
        if (fragment == null) {
            return;
        }
        final Context context = fragment.getContext();
        Intent intent = createIntent(context, product, imageType, mUrlImage);
        if (mImageFront!=null && fragment.getActivity() != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            ActivityOptionsCompat options = ActivityOptionsCompat.
                makeSceneTransitionAnimation(fragment.getActivity(), mImageFront,
                    fragment.getActivity().getString(R.string.product_transition));
            fragment.startActivityForResult(intent, ProductImageManagementActivity.REQUEST_EDIT_IMAGE, options.toBundle());
        } else {
            fragment.startActivityForResult(intent, ProductImageManagementActivity.REQUEST_EDIT_IMAGE);
        }
    }

    public static void openForUrl(Activity activity, Product product, ProductImageField imageType, String mUrlImage, View mImageFront) {
        if (activity == null) {
            return;
        }
        final Context context = activity.getBaseContext();
        Intent intent = createIntent(context, product, imageType, mUrlImage);
        if (mImageFront!=null && activity != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            ActivityOptionsCompat options = ActivityOptionsCompat.
                makeSceneTransitionAnimation(activity, mImageFront,
                    activity.getString(R.string.product_transition));
            activity.startActivityForResult(intent, ProductImageManagementActivity.REQUEST_EDIT_IMAGE, options.toBundle());
        } else {
            activity.startActivityForResult(intent, ProductImageManagementActivity.REQUEST_EDIT_IMAGE);
        }
    }

    private static Intent createIntent(Context context, Product product, ProductImageField imageType, String mUrlImage) {
        Intent intent = new Intent(context, ProductImageManagementActivity.class);
        String language = LocaleHelper.getLanguage(context);
        if (!product.isLanguageSupported(language) && StringUtils.isNotBlank(product.getLang())) {
            language = product.getLang();
        }
        intent.putExtras(ImageKeyHelper.createImageBundle(imageType, product, language, mUrlImage));
        return intent;
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/FullScreenActivityOpener.kt
package openfoodfacts.github.scrachx.openfood.features

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.view.View
import android.widget.Toast
import androidx.annotation.CheckResult
import androidx.core.app.ActivityOptionsCompat
import androidx.fragment.app.Fragment
import openfoodfacts.github.scrachx.openfood.R
import openfoodfacts.github.scrachx.openfood.features.images.manage.ImagesManageActivity
import openfoodfacts.github.scrachx.openfood.features.images.zoom.ImageZoomActivity
import openfoodfacts.github.scrachx.openfood.images.IMAGE_URL
import openfoodfacts.github.scrachx.openfood.images.ImageSize
import openfoodfacts.github.scrachx.openfood.images.createImageBundle
import openfoodfacts.github.scrachx.openfood.models.Product
import openfoodfacts.github.scrachx.openfood.models.ProductImageField
import openfoodfacts.github.scrachx.openfood.repositories.ProductRepository
import openfoodfacts.github.scrachx.openfood.utils.isAbsoluteUrl
import org.jetbrains.annotations.Contract

/**
 * Used to open fullscreen activity
 */
object FullScreenActivityOpener {

    suspend fun openForUrl(
        fragment: Fragment,
        client: ProductRepository,
        product: Product,
        imageType: ProductImageField,
        mUrlImage: String,
        mImageFront: View,
        language: String
    ) = openForUrl(fragment.requireActivity(), client, product, imageType, mUrlImage, mImageFront, language)

    suspend fun openForUrl(
        activity: Activity,
        client: ProductRepository,
        product: Product,
        imageType: ProductImageField,
        mUrlImage: String,
        mImageFront: View,
        language: String
    ) {
        // A new file added just now
        if (isAbsoluteUrl(mUrlImage)) {
            loadImageServerUrl(activity, client, product, imageType, mImageFront, language)
            return
        }
        startActivity(activity, mImageFront, createIntent(activity, product, imageType, mUrlImage, language))
    }

    private fun startActivity(activity: Activity, mImageFront: View?, intent: Intent) {
        val bundle = mImageFront?.let {
            ActivityOptionsCompat.makeSceneTransitionAnimation(
                activity,
                mImageFront,
                activity.getString(R.string.product_transition)
            ).toBundle()
        }
        activity.startActivityForResult(intent, ImagesManageActivity.REQUEST_EDIT_IMAGE, bundle)
    }

    fun openZoom(
        activity: Activity,
        imageUrl: String,
        imageView: View?
    ) = startActivity(activity, imageView, Intent(activity, ImageZoomActivity::class.java).apply {
        putExtra(IMAGE_URL, imageUrl)
    })

    @CheckResult
    @Contract(pure = true)
    private fun createIntent(
        context: Context,
        product: Product,
        imageType: ProductImageField,
        mUrlImage: String,
        language: String
    ): Intent {
        var productLanguage = language
        if (!product.isLanguageSupported(language) && product.lang.isNotBlank()) {
            productLanguage = product.lang
        }
        return Intent(context, ImagesManageActivity::class.java).apply {
            putExtras(createImageBundle(imageType, product, productLanguage, mUrlImage))
        }
    }

    private suspend fun loadImageServerUrl(
        activity: Activity,
        client: ProductRepository,
        product: Product,
        imageType: ProductImageField,
        mImageFront: View,
        language: String
    ) {
        client.getProductImages(product.code).product?.let { newProduct ->
            val imageUrl = newProduct.getSelectedImage(language, imageType, ImageSize.DISPLAY)
            if (!imageUrl.isNullOrBlank()) {
                openForUrl(activity, client, newProduct, imageType, imageUrl, mImageFront, language)
            } else {
                Toast.makeText(activity, R.string.cant_edit_image_not_yet_uploaded, Toast.LENGTH_LONG).show()
            }
        }
    }
}
